OneCommunityGlobal/HighestGoodNetworkApp
# Shiwani leaderboard time off indicator follow up

# Description
https://www.loom.com/share/6b34b79c77ae4c2ab408de5ea44e17c3?sid=95fa7a03-ad81-43d3-b9d4-941f994bac1d
This is a follow up PR of [#1805](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1805)

Follow up:
- Gray out name if the person is off for the current week
- If the user is currently on a time off then display the additional weeks off as + with a number next to it eg: +1 for 1 additional week excluding current week. If the person is taking only the current week off then + with a number would not be added.
- If the user has scheduled a future week off then a number without + would be added next to name in leaderboard which would indicate the number of weeks from now for the next time off to take place for that user. User name would not be in grey in this case.
- only roles above or equal to manager should be able to see the time off leaderboard indicator for other users.
- Add a tooltip to explain what the number next to name in leaderboard for each user represents.

## Related PRS (if any):
This frontend PR is related to the [#1805](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1805) and [#1722](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1722)

## Main changes explained:
- Added a new function called isUserGoingOnFutureTimeOff in timeOffRequestAction.js to calculate future weeks time off scheduled by users by calculating the closest start date from the time offs previously requested by a specific user and also only considering start dates greater than the current date. If such an entry is not present then it would return null.
- Modified getAllTimeOffRequests action in timeOffRequestAction.js by calling isUserGoingOnFutureTimeOff function and added calculation for additional weeks and future weeks to onVacation and futureTimeOff objects to be dispatched.
- Added future week constant to timeOffRequestConstants.js
- Added futureTimeOff to timeOffRequestReducer.js
- Removed previous additonal weeks time off calculation in leaderboard.jsx and used the weeks calculated in userOnTimeOff for additional weeks and used usersOnFutureTimeOff for future weeks.
- Added an array called allowedRoles which would have the roles that have access to see the leaderboard indicator for other users and checked if the logged in user's role matched any of them or if the logged in user's role is owner. Added condition for logged in user to see leaderboard indicator for their user name regardless of which role they belong to.
- Added tooltip to explain what the additional weeks and future weeks number next to name represents.
- Added usersOnFutureTimeOff to LeaderBoardContainer.jsx

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as any user who has Manage time off requests permission. If the user does not have this permission then log in as owner user and go to other links -> permission management -> manage user permissions -> select user name -> add 'Manage time off request' permission -> click submit -> log in as that user. If the user you are testing with already has this permission then feel free to skip the permission setting up step.
5. go to other links -> view profile -> schedule blue square
6. check for case 1: set the departure date as the previous sunday if the current day is not a sunday -> set the number of weeks to 1 -> click on submit -> close modal and click on dashboard. Check if the leaderboard indicator for your name has turned to grey without any +(number).
7. check for case 2: set the departure date as the previous sunday if the current day is not a sunday -> set the number of weeks to more than 1 -> click on submit -> close modal and click on dashboard. Check if the leaderboard indicator for your name has turned to grey with +(number) next to your name.
8. check for case 3: set the departure date to any upcoming sunday  -> set any number of weeks -> click on submit -> close modal and click on dashboard. Check if the leaderboard indicator for your name has not turned to grey and check if there is a number next to your name which indicates number of weeks from now for the next time off to take place.
9. check if hovering over the number next to name in leaderboard shows the tooltip. 
10. log in as any user lower to manager in the team example: volunteer and check if the leaderboard does not show time off indicators next to name for any other user than the logged in user's name in this case.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78906820/06029017-d7db-4865-9170-55956f3befce


## Note:
- If the user has scheduled a time off for the current week and if the user schedules a future time off then the leaderboard indicator would show only the current week time off indicator in this case. Once the current date passes the time off dates, the user will be able to see their future time off indicator. This is only for cases when a user has scheduled both a current and future week time off.
